hbase-client/src/main/java/org/apache/hadoop/hbase/MetaTableAccessor.java

```diff
@@ -2185,4 +2206,27 @@ private static Put addSequenceNum(Put p, long openSeqNum, int replicaId) throws
               .setValue(Bytes.toBytes(openSeqNum))
               .build());
   }
+
+  private static byte[] buildLocateRegionStartRow(TableName tableName, byte[] row,
+    RegionLocateType locateType) {
+    if (locateType.equals(RegionLocateType.BEFORE)) {
+      if (Bytes.equals(row, HConstants.EMPTY_END_ROW)) {
+        byte[] binaryTableName = tableName.getName();
+        return Arrays.copyOf(binaryTableName, binaryTableName.length + 1);
+      } else {
+        return createRegionName(tableName, row, ZEROES, false);
+      }
+    } else {
+      return createRegionName(tableName, row, NINES, false);
+    }
+  }
+
+  public static Scan createLocateRegionScan(TableName tableName, byte[] row,
+    RegionLocateType locateType, int prefetchLimit) {
```

Comment:
Why we need locateType now and not previously?

Reply:
Why is this not just a case of Scan forward or Scan backward?

Reply:
This is for abstraction, at this layer we do not care whether you are doing scan or what ever else, The only thing I care is how you want to locate a region through the given row, i.e, the direction.

And we have this logic, at least for two years after we introduced async client. If you are talking about the code before async client, the locating logic is placed everywhere in code, which is not good.

For supporting reverse scan, we have special logic to locate to the previous region in this class

https://github.com/apache/hbase/blob/branch-2/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ReversedScannerCallable.java

Reply:
hmmm... Went back and saw the locateInMeta.  This replicates the head of that method? It should use this?

Reply:
We could be better orginzation of these methods in follow on patches, i.e, moving this method to other places to not mess up the MetaTableAccessor. I do not have strong feelings that this method must be put here.